src/coreclr/jit/lclmorph.cpp

```diff
@@ -2426,3 +2426,520 @@ void Compiler::fgExposeLocalsInBitVec(BitVec_ValArg_T bitVec)
+        return;
+    }
+
+    LCLMasksWeight weight = {0, 0};
+    bool           found  = weightsTable->Lookup(lclVar->GetLclNum(), &weight);
+
+    // If there no entry, then the var does not have a local store.
+    if (!found)
+    {
+        JITDUMP("Local Var V%02d at [%06u] is not stored to local.\n", lclVar->GetLclNum(), dspTreeID(lclVar));
+        return;
+    }
+
+    if (!weight.MaskConversionsDominate())
+    {
+        JITDUMP("Local Var V%02d at [%06u] will not be converted. Weighting {%d, %d}\n", lclVar->GetLclNum(),
+                dspTreeID(lclVar), weight.storeWeight, weight.varWeight);
+        return;
+    }
+
+    JITDUMP("Local Var V%02d at [%06u] will be converted. Weighting {%d, %d}\n", lclVar->GetLclNum(), dspTreeID(lclVar),
+            weight.storeWeight, weight.varWeight);
+
+    // Remove or add a mask conversion/
+    LCLMasksUpdateLCLVarVisitor ev(this, lclVar->GetLclNum(), stmt, weight.simdBaseJitType, weight.simdSize);
+    GenTree*                    root = stmt->GetRootNode();
+    ev.WalkTree(&root, nullptr);
+}
+
+#endif // TARGET_ARM64
+
+//------------------------------------------------------------------------
+// optLCLMasks: Allow locals to be of Mask type
+//
+// At the C# level, Masks share the same type as a Vector. It's possible for the same
+// variable to be used as a mask or vector. Any APIs that return a mask must first convert
+// the value to a vector before storing it to a variable. Any uses of a variable as a mask
+// must first convert from vector before using it. In many cases this creates unnecessary
+// conversions. For variables that live outside the scope of the current method then the
+// conversions are required to ensure correctness. However, for local variables where the
+// scope is local to the current method, then it is possible to keep the value as a mask,
+// by updating all definitions and uses.
+//
+// In the common case it is expected that uses of masks are consistent - once a variable is
+// created as a mask it will continue to be used and updated as a mask.
+//
+// In the uncommon case, a variable may be created in one type, used as another and/or
+// updated to a different type.
+//
+// For example (the conversion is implicit)
+//   vector<int> x = _ConvertMaskToVector_(CreateMask());
+//   x = Add(x, y);
+//
+// To account for this, this pass uses a weighting. For each variable, count the count the
+// number of definitions with a convert from mask minus the number of definitions without a
+// convert. Then do the same for each use. If both totals for a variable are positive, then
+// convert every definition and use to use a mask instead of a vector.
+//
+// This weighting does not account for:
+// * Loops. Uses/definitions inside a loop will account for more real weight than any outside
+// a loop. This is not expected to be an issue as in most cases when using loops, a variable
+// will be set once outside a loop then used/defined multiple times inside a loop.
```

Comment:
It wouldn't be that hard to take this into account... you can use the block weight to do it. Usually for heuristics like this it's nice to have a proper model of "how many cycles and bytes does a conversion cost" and then phrase the heuristic as (sum over BBs : bb->getBBWeight() * (numConvertedToMask - numConvertedFromVector)) > 0, additionally taking size into account as a tie breaker. For example, that's what the heuristic for physical promotion looks like:

https://github.com/dotnet/runtime/blob/7c2f0041be9068fb641f53dc830606c95aab2488/src/coreclr/jit/promotion.cpp#L840-L875

If you think it's unnecessary then I am ok with this heuristic too.

Reply:
> The problem is that at this point the bbWeights are all 100, regardless of any looping.

That's probably because you are running your tests with tiered compilation disabled. This is going to disable dynamic PGO which is where we get accurate block weights from. The vast majority of users are going to have dynamic PGO enabled, so you don't need to worry about this.

> I'm not sure if it is safe to move this phase after weights have been calculated? Also, I'm using LocalsTreeList() which requires fgNodeThreading == AllLocals. Any suggestions?

The "Set block weights" phase is expected to be removed soon. The replacement is PGO synthesis, which runs during the "Profile incorporation" phase, way before your phase. It is not enabled by default today since we normally get the data from dynamic PGO which is more accurate. However, for scenarios like NAOT or crossgen2 we are still expecting to enable PGO synthesis at some point since those don't have dynamic PGO available.

You can test with PGO synthesis enabled by setting `DOTNET_JitSynthesizeCounts=1`, but the more realistic test would be to enable tiered compilation and make sure that your benchmarks are tiered up in the real-world way.

In other words, I would keep things exactly as they are, but you may need to change your testing methodology to verify that things are working as expected.

Reply:
Here's an example regression test for a problem that required dynamic PGO to reproduce, and that thus needs to tier up the benchmark: https://github.com/dotnet/runtime/blob/ffcef3dec5f16289a8f71f332091266529288d5c/src/tests/JIT/Regression/JitBlue/Runtime_95315/Runtime_95315.cs

Reply:
> In other words, I would keep things exactly as they are, but you may need to change your testing methodology to verify that things are working as expected.

Got it, thanks!